README.txt
Ace Editor Module ( https://drupal.org/project/ace_editor )
======================
by Max Nylin, max@articstudios.se
by Olivier GIRARD, girardol182@gmail.com
by interdruper, https://www.drupal.org/u/interdruper


Description
===========

AceEditor is a code editor written in JavaScript, allowing you to edit HTML,
PHP and JavaScript (and more). It provides syntax highlighting, proper
indentation, keyboard shortcuts, find and replace (including regular
expressions).

This module integrates the Ace editor into Drupal's node/block edit forms,
for editing raw HTML/PHP/JavaScript (with more) in a familiar way.
It also provides a display formatter, along with a text filter and
an API to embed and show code snippets in your content.


Installation
============

a) Using Drush (recommended from 7.x-1.2):
     Just run 'drush en ace_editor'. The module will be downloaded and enabled.
     While enabling, you will be asked to pick the desired Ace library release
     to be installed in sites/all/libraries. Drush will download and extract
     the Ace library using the command 'drush dl-ace'.

     An new "Ace Editor" filter format is added. It uses the Ace editor
     for editing content. You can enable the Ace editor also for any other
     filter formats at admin/config/content/ace-editor.

b) Manually:
    1. Make sure that Libraries 7.x-2.x module is installed and enabled.
    2. Download the latest version of the Ace Editor at
       https://github.com/ajaxorg/ace-builds/ or directly
       via https://github.com/ajaxorg/ace-builds/archive/master.zip
       Do not use a version < 1.0.0.
    3. Extract and place the contents of the zip file under
       sites/all/libraries so that ace.js is located at
       sites/all/libraries/ace/src/ace.js. If you want to use any other src
       (minified, noconflict...), make sure that you rename the folder as "src"
       for allowing the module to find it.
    4. Download, extract and copy the ace_editor module to your
       sites/all/modules or sites/all/modules/contrib directory.
    5. Enable the "Ace HTML Editor" module on your Drupal Modules page,
       under the Administration heading. An example "Ace Editor" filter format
       is added that uses the Ace editor for editing content. You can enable the
       Ace editor for other filter formats at admin/config/content/ace-editor.

Uninstallation
==============

The module adds a filter format named 'Ace Editor' on installation. This filter
format is not disabled when the module is uninstalled, to preserve any content
saved using it. If you are sure that there is no valuable content in your
database saved under the 'Ace Editor' filter format, you can manually disable
the filter format at admin/config/content/formats.

Drush support
=============

The module offers the Drush command 'dl-ace', for easily install and/or upgrade
the Ace library in sites/all/libraries. Please run 'drush dl-ace --help' for
details.

Dependencies
============

Libraries API module (https://www.drupal.org/project/libraries)


Features
========

Edit HTML and PHP in your nodes and blocks like a pro
-----------------------------------------------------

Go to admin/config/content/ace-editor and configure the module for
node/block editing. Then head over to a block or node containing a textarea
with the correct text format and hack away!

Autocompletion
--------------

Press Ctrl+Space to use the autocomplete option while coding.

Display fields using syntax highlighting
----------------------------------------------

Manage the display of any text area fields attached to a node and select
the "Code syntax highlighting" format. This outputs the content of the field
as a ready-only editor, with syntax highlighting in your node view using
the selected options.


Use syntax hilighting from your template files
-----------------------------------------------------

You can use the ace_editor_add($content, $settings) function to add
syntax-highlighting code display anywhere in your template files.
An optional array contains settings as shown below.


Embed code snippets in the body of your nodes or blocks
---------------------------------------------------------------

Add the syntax highlighting filter to any of your text formats. The module
displays text inside an <ace> tag as code using the custom formatting options
specified as attributes to the <ace> tag.

You can override the default options by adding attributes to the <ace> tag:

Here are the possible values:

  theme
    clouds = Clouds
    clouds_midnight = Clouds Midnight
    cobalt = Cobalt
    crimson_editor = Crimson Editor
    dawn = Dawn
    idle_fingers = Idle Fingers
    kr_theme = krTheme
    merbivore = Merbivore
    merbivore_soft = Merbivore Soft
    mono_industrial = Mono Industrial
    monokai = Monokai
    pastel_on_dark = Pastel on dark
    solarized_dark = Solarized Dark
    solarized_light = Solarized Light
    textmate = TextMate
    twilight = Twilight
    tomorrow = Tomorrow
    vibrant_ink = Vibrant Ink

    ... and any other theme-*.js file available in sites\all\libraries\ace\src

  syntax
    c_cpp = C/C++
    clojure = Clojure
    coffee = CoffeeScript
    csharp = C#
    css = CSS
    groovy = Groovy
    html = HTML
    java = Java
    javascript = JavaScript
    json = JSON
    ocaml = OCaml
    perl = Perl
    php = PHP
    python = Python
    scala = Scala
    scss = SCSS
    ruby = Ruby
    svg = SVG
    textile = Textile
    xml = XML

    ... and any other mode-*.js file available in sites\all\libraries\ace\src

  height
    auto, 300px, 75% etc.

  width
    100%, 600px etc.

  font-size
    All compatible CSS values for font-size

  line-numbers
    1 or 0 (on/off)

  print-margin
    1 or 0 (on/off)

  invisibles
    1 or 0 (on/off)

Examples:
      <ace theme="textmate" height="200px" font-size="12pt" print-margin="1">
      <ace theme="twilight" syntax="php" height="200px" width="50%">
      <ace height="100px" width="100%" invisibles="1">


CHANGELOG.txt
7e4f5fa (HEAD, 7.x-1.x) Issue #2737727 by aburke626: Code Style Cleanup
f69c497 (origin/HEAD, origin/7.x-1.x) Issue #2689757 by interdruper,abeal: The editor doesn't work with node edit form while good for syntax highlight display (read only)
12b3f63 Many minor styling fixes from code review.
0d67202 Issue #2693631: Ace editor assets are not properly loaded if Libraries is not enabled before installation
77f6904 Fixed: Notice: Undefined index: ace in ace_editor_install() (line 34 of ace_editor.install
62bd0fe (tag: 7.x-1.7) Issue #2691557: [drush] call to undefined function libraries_get_path()
e39fcbd Issue #2691575 by Ruichao: [drush] Mod and theme list not loaded via drush en
0c6aa8e Fixing explanation in the admin form.
d0350fb CHANGELOG.txt updated.
633f202 Issue #2683745: Views - 'Global: Text area' support (in Header, Footer, No Results Behavior).
7335bc5 (tag: 7.x-1.6) Issue #2645864: Autocomplete isn'\''t implemented (fixed).
20a1176 Issue #2645806 by OnkelTem,interdruper: [drush] Correct version fetching
8660548 Issue #2659476 by pobster, interdruper: Worker .js files not found, console error, syntax highlight doesn't work
c4c6337 Added option inside the editor to set the tabsize.
46bb58e Issue #2645880 by OnkelTem,interdruper: Add full list of Themes and Modes
08d84de Do not call drush_user_abort() if user selects not to download the library when installing with drush.
4808dca Issue #2645848 by onkeltem: Uncaught TypeError: Cannot read property form of undefined.
e642546 Issue #2643228 by interdruper: adding code folding option.
1dd13f4 (tag: 7.x-1.5) Issue #2409583 by jarviss,interdruper: add Ace Editor to Devep php execute form.
83e5b80 Issue #2641022 by interdruper: Content changes not saved.
d48b735 CHANGELOG.txt updated.
3e6cd94 (tag: 7.x-1.4) Issue #2255597 by interdruper: Make it work with jQuery 1.10 and above
82210ac Issue #2255597 by interdruper: Make it work with jQuery 1.10 and above
ef6ab82 Fixes in drush command 'dl-ace'.
a079406 README.txt updated
76dd09a (tag: 7.x-1.3) Revert "Issue #2255597 by interdruper, alex-bukach: Make it work with jQuery 1.10 and above"
a020316 (tag: 7.x-1.2) Issue #2255597 by interdruper, alex-bukach: Make it work with jQuery 1.10 and above
26a853f Issue #2426307 by interdruper, Quty: Saving field with an ace format fails the first time
7542015 Issue #2426307 by interdruper, Quty: Saving field with an ace format fails the first time
7b9f1a5 Issue #2625650 by Infanty, interdruper: Fix PHP syntax highlighting: missing base_path in library source path.
ab41d17 Code styling: replacing tabs by spaces in all source files.
418888c Issue #2330383 by mkhamash,interdruper: Multi-site libraries path.
916e6ab Show all the installed themes in the theme list.
4780cf5 CHANGELOG.txt updated.
544afe2 Issue #414278 by interdruper: Display formatter that outputs the content in an ace editor: fixed minor issues, available options reviewed.
c1709d6 Using libraries_detect() for checking if Ace Library is properly installed.
64417be Added hook_libraries_info().
755ac3e Removing some commented code from ace_editor.drush.inc.
87f37e7 Added Drush command 'dl-ace' for easily download and install the Ace library (some fixes).
e381ca0 Added Drush command 'dl-ace' for easily download and install the Ace library.
7ef4102 README.txt: fixing typos.
c4316ed Do not create but just enable the 'Ace Editor' filter format if the format has been previously disabled.
b26df89 README.txt updated with instructions for removing Ace Editor filter format after unsintalling the module.
2fe66d4 Revert "Disable Ace Editor filter format if module is uninstalled." Content saved with 'Ace Editor' filter format will not be shown. This reverts commit 7cdc3cb8019f1ff8b74e8edc32913b031b61a02f.
7cdc3cb Disable Ace Editor filter format if module is uninstalled.
481049a Added hook_requirements() to provide a library status message on the Status Report.
b47c8fb Issue #2330383 by torotil: Multi-site libraries path.
9f50437 Issue #2468489 by torotil: Use drupal_html_id instead of custom code.
fc0d2cb Issue #2431371 by interdruper: Option inside the editor to dinamically enable or disable auto-completion.
b1cd37b Issue #1420290 by interdruper: Added autowrap and line highlighting options to admin panel.
d44843f Issue #2270799 by Alex Bukach: Not working when library is not in sites/all/libraries. Added dependency from Libraries API module.
7dac151 Added CHANGELOG.txt file.
22398f8 Issue #2287147 by tonylegrone: Add new custom permission for the config page of the module
b0fb30d (tag: 7.x-1.1) #2230725 by Growiel,tempo22: Update the README to specify that we need to use a version of Ace > 1.0.0 and update the download links to reflect that
b41f13d Issue #2227443 by Growiel: Editor is now able to calculate its own height
a9c2e91 (tag: 7.x-1.0) Small tweaks to the README and fix a comment in ace_editor_install. We're ready for a stable release.
f553f37 Issue #1425558 by Amitaibu: change the way we include JS files to prevent conflicts
ab94907 Issue #1933320 by Growiel: remplace .html by .text to allow the use of <script> tags
66efca2 Issue #2090921 by tjerah: Don't preprocess the JS to make them work with aggregated. Issue #2192611 by growiel: Don't remove the first letter and allow multiple instance of the exact same code
cd6e878 (tag: 7.x-1.0-alpha12) FIxed a bug where the settings for the API didn't apply.
f188389 Updated README.
108b540 Changed attributes on the <pre> tag to reflect css attributes better to avoid confusion.
0267269 Updated README.
5fe89f3 Updated README.
0a497de (tag: 7.x-1.0-alpha11) Added documentation.
4ad9b3e No longer uses localStorage if browser does not support it.
148ebb3 Issue #1417680 by nylin: Resize editors and make changes persistent.
64efcde Issue #1415160 by nylin: Activating Ace editor reveals summary field.
457d461 (tag: 7.x-1.0-alpha10) Added all themes and modes to all places the editor could be used.

js/ace_edit.js
(function($) {

  Drupal.behaviors.ace_editor = {
    attach: function(context, settings) {
		
		/**
		 * If the user selected a text format configured to be used with the editor,
		 * show it, else show the default textarea.
		 */
		function setAceState($textArea) {
			
			// Top Level, for traversing.
			var $textFormatWrapper = $textArea.parents('div.text-format-wrapper:first');
			// The container that will hold the <pre>.
			var $formItem = $textFormatWrapper.find('div.form-item[class*="-value"]:first');
			// The select list for chosing the text format that will be used.
			var $filterSelector = $textFormatWrapper.find('select.filter-list');

			// If the text format's name is ace_editor then we'll add the editor.
			if ($.inArray($filterSelector.val(), Drupal.settings.ace_editor.text_formats) != -1) {
				
				var aceEditor, $ace_editor;
				
				// Check to see if the pre element exists, if not, create it.
				if (!$formItem.find('div.ace-editor').length) {
					// Create the tag and add default styling to it.
					
					$ace_editor = $('<div class="ace-editor"></div>');
					//$ace_editor_container.append(get_editor_controls());
					var $pre = $('<pre id="' + $textArea.attr('id') + '-aced"></pre>');
					$pre.css({
						'position': 'relative',
						'height': '600px',
						'border': '1px solid #ccc',
						'margin-bottom': '0px',
						'font-size': '10pt'
					});
					
					$ace_editor.append($pre);
					$formItem.append($ace_editor);
					
					// Ace it! TODO: Add settings page to let the user configure this.
					aceEditor = ace.edit($pre.attr('id'));
				    aceEditor.setTheme("ace/theme/twilight");
					var HTMLMode = require("ace/mode/html").Mode;
				    aceEditor.getSession().setMode(new HTMLMode());
					aceEditor.setShowPrintMargin(false);
					
					// Store the editor instance to the pre element for later use.
					$ace_editor.data('ace-editor', aceEditor);
					
				} else {
					
					// Find the pre elements.
					$ace_editor = $formItem.find('div.ace-editor');
					
					// The pre tag exists, and so does the editor instance.
					aceEditor = $ace_editor.data('ace-editor');
				}
				
				// Add the content of the field to the editors current session.
				aceEditor.getSession().setValue($textArea.val());
				
				// Show the editor's pre and hide the textarea.
				$ace_editor.show();
				$formItem.find('div.form-textarea-wrapper').hide();
				
			} else {
				
				// Fins the pre element.
				var $ace_editor = $formItem.find('div.ace-editor');
				
				// Fetch the Ace Editor instance from the pre.
				var aceEditor = $ace_editor.data('ace-editor');
				
				// Set the text of the textarea to reflect the changes in the Ace Editor.
				$textArea.val(aceEditor.getSession().getValue());

				// Hide the Ace Editor and show the textarea.
				$ace_editor.hide();
				$formItem.find('div.form-textarea-wrapper').show();
			}
		}
		
		/**
		 *	Returns the editor controls.
		 */
		function get_editor_controls() {
			$controls = $('<div class="ace-editor-controls"></div>');
			$controls.append(<div></div>);
		}
		
		/**
		 * Bind the change event to all text format select lists.
		 */
		$('div.text-format-wrapper fieldset.filter-wrapper select.filter-list').live('change', function(e) {
			var $textArea = $(this).parents('div.text-format-wrapper:first').find('div.form-type-textarea[class*="-value"] textarea');
			setAceState($textArea);
		});
	
		/**
		 * Transfer over the html of the editor to the correct field.
		 */
		$('#edit-submit').click(function() {
			$('div.form-item.form-type-textarea').each(function (i) {
				var $pre = $(this).find('pre.ace_editor');
				if ($pre.length) {
					var filterListValue = $(this).siblings('fieldset.filter-wrapper:first').find('select.filter-list').val();
					if (filterListValue == 'ace_editor') {
						var $textArea = $(this).find('textarea');
						var aceEditor = $pre.data('ace-editor');
						var html = aceEditor.getSession().getValue();
						$textArea.val(html);
					}
				}
			});
		});
	
		/**
		 *  
		 */
		// The change event of the text format select lists seens to be dispatched on document ready...
		/*
		$('div.form-type-textarea[class*="-value"] textarea').each(function(i) {
			// Is in a text format wrapper.
			if ($(this).parents('div.text-format-wrapper').length) {
				//setAceState($(this));
			}
		});
		*/
		
    }
  };

})(jQuery);


js/ace_editor.field_format.js
(function($) {

  Drupal.behaviors.ace_editor_field_format = {
	attach: function(context, settings) {
		
		var fieldFormatSettings = Drupal.settings.ace_editor.field_format_settings;
		
		var SyntaxMode = require("ace/mode/" + fieldFormatSettings['syntax']).Mode;
		var themePath = "ace/theme/" + fieldFormatSettings['theme'];
		$(Drupal.settings.ace_editor.field_format_instances).each(function(i) {
			var $pre = $('#' + this['id']);
			var html = this['content'];
			
			var editor_instance = ace.edit(this['id']);
			editor_instance.getSession().setMode(new SyntaxMode());
			editor_instance.setTheme(themePath);
			editor_instance.renderer.setShowGutter(fieldFormatSettings['line_numbers']);
			editor_instance.setShowPrintMargin(fieldFormatSettings['print_margin']);
			editor_instance.setShowInvisibles(fieldFormatSettings['invisibles']);
			editor_instance.setReadOnly(true);
			$pre.css({
				'font-size': formatFilterSettings['font_size'],
				'height': formatFilterSettings['field_height'],
				'width': formatFilterSettings['field_width']
			});
			
			editor_instance.getSession().setValue(html);
		});
		
		
	}
  };

})(jQuery);


js/ace_editor.format_filter.js
(function($) {

  Drupal.behaviors.ace_editor_field_format = {
	attach: function(context, settings) {
		
		var formatFilterSettings = Drupal.settings.ace_editor.format_filter_settings;
		
		var SyntaxMode = require("ace/mode/" + formatFilterSettings['syntax']).Mode;
		var themePath = "ace/theme/" + formatFilterSettings['theme'];
		$(Drupal.settings.ace_editor.format_filter_instances).each(function(i) {
			var $pre = $('#' + this['id']);
			var html = this['content'];
			
			var editor_instance = ace.edit(this['id']);
			editor_instance.getSession().setMode(new SyntaxMode());
			editor_instance.setTheme(themePath);
			editor_instance.renderer.setShowGutter(formatFilterSettings['line_numbers']);
			editor_instance.setShowPrintMargin(formatFilterSettings['print_margin']);
			editor_instance.setShowInvisibles(formatFilterSettings['invisibles']);
			editor_instance.setReadOnly(true);
			$pre.css({
				'font-size': formatFilterSettings['font_size'],
				'height': formatFilterSettings['field_height'],
				'width': formatFilterSettings['field_width']
			});
			
			// Set the height of the editor if field_height is set to auto.
			/*
			if (formatFilterSettings['field_height'] == 'auto') {
				editor_instance.getSession().on('change', function(e) {
					var cursorHeight = $pre.find('div.ace_cursor:first').height();
					//alert(cursorHeight);
					$pre.css('height', (editor_instance.getSession().getValue().split('\n').length + 2) * cursorHeight);
				});
			}
			*/
			
			editor_instance.getSession().setValue(html);
		});
		
		
	}
  };

})(jQuery);


js/ace_editor.admin.js
(function($) {

  Drupal.behaviors.ace_editor_admin = {
  attach: function(context, settings) {

    // Contains the settings for the editors.
    var editorSettings;
    var formIdentifier;

    /**
     * If the user selected a text format configured to be used with the editor,
     * show it, else show the default textarea.
     */
    function acifyWrapper($textFormatWrapper) {
      // The select list for chosing the text format that will be used.
      var $filterSelector = $textFormatWrapper.find('select.filter-list');

      // Checks if the currently selected filter is one that uses the editor.
      if ($.inArray($filterSelector.val(), Drupal.settings.ace_editor.admin.text_formats) != -1) {

        editorsJustAdded = false;

        /**
         * If the settings has not been set I will copy them into a new object.
         * if I don't the settings will be duplicated every time I add a
         * textarea on fields that supports multiple values.
         */
        if (!editorSettings) {
          editorSettings = $.extend({}, Drupal.settings.ace_editor.admin);
          formIdentifier = $textFormatWrapper.parents('form:first').find('input[name="ace_editor_identifier"]').val();
        }

        // Check to see if the editor has been added yet.
        if ($textFormatWrapper.find('div.ace-editor-container').length !=
          $textFormatWrapper.find('div.form-item.form-type-textarea').length) {

          // Init the array that will hold all editor elements and instances
          // inside of each text format wrapper.
          var editors = new Array();

          // Iterate through all textarea containers that and attach the editor.
          $('div.form-item.form-type-textarea:visible', $textFormatWrapper).each(function(i) {
            var $form_item = $(this);
            var editor_instance;

            if (!$form_item.find('pre').length) {
              // Create the editors container.
              var $ace_editor_container = $('<div class="ace-editor-container"></div>');
              // Put the different parts together.
              var $pre = $('<pre id="' + $(this).find('textarea').attr('id') + '-aced"></pre>');
              $ace_editor_container.append($pre);

              // Get the stored height for the fields if it exists.
              var sizeId = formIdentifier + '_' + $form_item.find('textArea').attr('id').replace(new RegExp('-', 'g'), '_')
              var storedHeight = localStorageGet('ace_editor_' + sizeId + '_height');
              var height = storedHeight ? storedHeight : null;

              // Only append controlls to the main form item.
              var $controls;
              if ($(this).attr('class').indexOf("-value") != -1) {

                $pre.css({'height': height ? height : '500px'});
                $controls = get_editor_controls();
                $ace_editor_container.append($controls);
                $(this).append($ace_editor_container);
              } else {
                // It's a summary form item.
                $pre.css({'height': height ? height : '200px', 'border-bottom': '1px solid #CCC'});
                $(this).find('div.form-textarea-wrapper').after($ace_editor_container);
              }

              // Initialize the editor and set the correct options.
              editor_instance = ace.edit($pre.attr('id'));
              ace.config.set('basePath', editorSettings['ace_src_dir']);
              editor_instance.setTheme("ace/theme/" + editorSettings.theme);
              var HTMLMode = ace.require("ace/mode/html").Mode;
              editor_instance.getSession().setMode(new HTMLMode());
              editor_instance.setShowPrintMargin(editorSettings['printmargin']);
              editor_instance.getSession().setUseWrapMode(editorSettings['autowrap']);
              editor_instance.setHighlightActiveLine(editorSettings['linehighlighting']);
              editor_instance.getSession().setFoldStyle(editorSettings['codefolding']);
              editor_instance.renderer.setHScrollBarAlwaysVisible(false);
              $pre.css('font-size', editorSettings['fontsize']);
              $pre.data('editor_instance', editor_instance);
              // Enable options and snippets.
              ace.require("ace/ext/language_tools");
              editor_instance.setOptions({
                enableBasicAutocompletion: true,
                enableLiveAutocompletion: false,
                tabSize: editorSettings['tabsize'],
                useSoftTabs: true
              });

              // Adds handles to the editors for vertical resizing.
              $pre.resizable({
                handles: 's',
                resize: function(event, ui) {
                  $pre.width('100%');
                  editor_instance.resize();
                },
                stop: function(event, ui) {
                  var sizeId = formIdentifier + '_' + $form_item.find('textArea').attr('id').replace(new RegExp('-', 'g'), '_')
                  localStorageSet('ace_editor_' + sizeId + '_height', $pre.height());
                }
              });

              // Add event listeners.
              editor_instance.getSession().on('change', function(editor) {
                editorContentChange($form_item, editorObject);
              });

              editorsJustAdded = true;
            } else {
              $pre = $form_item.find('pre');
              editor_instance = $pre.data('editor_instance');
            }

            var editorObject = {
              editor: editor_instance,
              element: $pre
            };
            editors.push(editorObject);
          });

          // Store the newly created editor instances for later use.
          $textFormatWrapper.data('ace-editors', editors);

          // Set control states.
          $textFormatWrapper.find('div.ace-editor-controls').each(function(i) {
            var $controls = $(this);

            var $textArea = $textFormatWrapper.find('textarea');
            var textAreaID = formIdentifier + '_' + $textArea.attr('id').replace(new RegExp('-', 'g'), '_');

            var showHidden = localStorageGet('ace_editor_' + textAreaID + '_show_hidden');
            var lineNumbers = localStorageGet('ace_editor_' + textAreaID + '_show_line_numbers');
            var mode = localStorageGet('ace_editor_' + textAreaID + '_mode');

            $controls.find('input.show_hidden').attr('checked', (showHidden == 1) ? true : false);
            $controls.find('input.show_line_numbers').attr('checked', (lineNumbers == 1) ? true : false);
            $controls.find('select.mode').val(mode ? mode : 'html');

            // Trigger controls.
            $controls.find('div.control input, div.control select').trigger('change');
          });
        }

        // Get all of the editor objects for the text format wrapper.
        var editorObjects;
        if (editorsJustAdded) {
          editorObjects = editors;
        } else {
          editorObjects = $textFormatWrapper.data('ace-editors');
        }

        // Show the editor containers and hide the textarea containers.
        $(editorObjects).each(function(i) {
          var $formItem = $(this["element"]).parents('div.form-item:first');
          var $textAreaWrapper = $formItem.find('div.form-textarea-wrapper');

          if ($textAreaWrapper.is(":visible")) {
            var $editorContainer = $formItem.find('div.ace-editor-container');

            // Add the content of the field to the editors current session.
            this["editor"].getSession().setValue($textAreaWrapper.find('textarea').val());

            $textAreaWrapper.hide();
            $editorContainer.show();
          }
        });

      } else {
        // Show the textarea.
        var editorObjects = $textFormatWrapper.data('ace-editors');
        if (editorObjects) {

          // Only switch of editors are shown.
          var $firstElement = $(editorObjects[0]["element"]);
          if ($firstElement.is(':visible') || !$firstElement.parent().parent().is(':visible')) {
            $(editorObjects).each(function(i) {
              var editorObj = editorObjects[i];

              var $formItem = $(editorObj["element"]).parents('div.form-item:first');
              var $textAreaWrapper = $formItem.find('div.form-textarea-wrapper');
              var $editorContainer = $formItem.find('div.ace-editor-container');

              // Transfer content from editor to textarea.
              $textAreaWrapper.find('textarea').val(editorObj['editor'].getSession().getValue())

              // Hide the editor containers and show the textarea containers.
              $textAreaWrapper.show();
              $editorContainer.hide();
            });
          }
        }

      }
    }

    /**
    * Bind the change event to all text format select lists.
    */
    var toAcify = 'div.text-format-wrapper fieldset.filter-wrapper select.filter-list';

    $(toAcify).change(function(e) {
      var $textFormatWrapper = $(this).parents('div.text-format-wrapper:first');
      acifyWrapper($textFormatWrapper);
    });

    /**
     * Update the editors to reflect the toggled option.
     */

    // Customize this handler depending on the installed JQuery version,
    // to ensure compatibility from JQuery 1.4 to 1.9+. Issue #2255597.
    var iCanUseOn = !!$.fn.on;
    if (iCanUseOn) {
      var func = 'on';
      var selections = 'div.text-format-wrapper div.control input, div.text-format-wrapper div.control select';
      var container = 'div#content';
    }
    else {
      var func = 'live';
      var selections = 'div.text-format-wrapper div.control input, div.text-format-wrapper div.control select';
      var container = selections;
    }

    $(container)[func]('change', selections, function(e) {

      var $control = $(this);
      var $textFormatWrapper = $(this).parents('div.text-format-wrapper:first');
      var editorObjects = $textFormatWrapper.data('ace-editors');
      var $textArea = $textFormatWrapper.find('textarea');
      var textAreaID = formIdentifier + '_' + $textArea.attr('id').replace(new RegExp('-', 'g'), '_');

      var checked;
      if ($control.attr('type') == 'checkbox') {
        checked = $(this).is(':checked') ? 1 : 0;
      }

      if ($control.attr('name') == 'mode') {
        if ($control.val() == 'html') {
          // Apply settings to all editors.
          $(editorObjects).each(function(i) {
            var HTMLMode = ace.require("ace/mode/html").Mode;
            this['editor'].getSession().setMode(new HTMLMode());
          });
        } else {
          $.getScript(editorSettings['ace_src_dir'] + 'mode-' + $control.val() + '.js', function(data, textStatus) {
            $(editorObjects).each(function(i) {
              var Mode = ace.require("ace/mode/" + $control.val()).Mode;
              this['editor'].getSession().setMode(new Mode());
            });
          });
        }
      } else {
        // Apply settings to all editors.
        $(editorObjects).each(function(i) {
          switch ($control.attr('name')) {
            case "show_hidden":
              this['editor'].setShowInvisibles(checked);
              break;

            case "show_line_numbers":
              this['editor'].renderer.setShowGutter(checked);
              break;

            case "autocomplete":
              this['editor'].setBehavioursEnabled(checked);
              break;
          }
        });
      }

      // Save settings 'till next time.
      switch ($control.attr('name')) {
        case "show_hidden":
          localStorageSet('ace_editor_' + textAreaID + '_show_hidden', checked);
          break;

        case "show_line_numbers":
          localStorageSet('ace_editor_' + textAreaID + '_show_line_numbers', checked);
          break;

        case "autocomplete":
          localStorageSet('ace_editor_' + textAreaID + '_autocomplete', checked);
          break;

        case "mode":
          localStorageSet('ace_editor_' + textAreaID + '_mode', $control.val());
          break;
      }
    });

    /**
     * Add the editor to the summary field.
     * Seems that drupal blocks the use of click here, so we need to use
     * mouseup with a small delay to get it to work.
     */
    $('div#content').mouseup('a.link-edit-summary', function(e) {
      window.setTimeout(function() {
        acifyWrapper($('a.link-edit-summary').parents('div.text-format-wrapper:first'));
      }, 10);
    });

    /**
     * The content of the editor has changed, update the span showing line
     * numbers. Also transfer the content of the editors to their related
     * textareas.
     */
    function editorContentChange($form_item, editorObject) {
      $(editorObject['element']).parents('div.ace-editor-container:first').find('div.ace-editor-controls span.num-lines').text(editorObject['editor'].getSession().getValue().split("\n").length + " lines");

      var $textarea = $form_item.find('textarea');
      var val = editorObject['editor'].getSession().getValue();
      $textarea[0].value = val;
    }

    /**
    *  Returns the controls for an editor.
    */
    function get_editor_controls() {
      $controls = $('<div class="ace-editor-controls"></div>');
      $controls.append('<div class="control"><input type="checkbox" name="show_hidden" class="show_hidden" checked>' +
        '<label>Invisibles</label></div>');
      $controls.append('<div class="control"><input type="checkbox" name="show_line_numbers" class="show_line_numbers" checked>' +
        '<label>Line numbers (show errors)</label></div>');
      $controls.append('<div class="control"><input type="checkbox" name="autocomplete" class="ace_autocomplete" checked>' +
        '<label>Auto-pairing</label></div>');
      $modes_select = $('<div class="control"><select name="mode" class="mode"></select></div>');
      $.each(editorSettings['available_modes'], function(key, value) {
        var selected = (key == 'html') ? ' selected' : '';
        $('select', $modes_select).append('<option value="' + key + '"' + selected + '>' + value + '</option>');
      });
      $controls.append($modes_select);

      $controls.append('<div class="info"><span class="num-lines">1 lines</span><a class="key-bindings" target="_blank" href="https://github.com/ajaxorg/ace/wiki/Default-Keyboard-Shortcuts">Show key bindings</a></div>');

      return $controls;
    }

    /**
     * Saves data to localStorage if available.
     */
    function localStorageSet($key, $value) {
      if (localStorageAvailable()) {
        localStorage[$key] = $value;
      }
    }

    /**
     * Gets data from localStorage if available.
     */
    function localStorageGet($key) {
      if (localStorageAvailable()) {
        return localStorage[$key];
      }
      return null
    }

    /**
     *
     */
    function localStorageAvailable() {
      return 'localStorage' in window && window['localStorage'] !== null;
    }

    /**
     * PHP consoles handling (Devel, Views PHP if available).
     */
    // @TODO: Views PHP: textarea#edit-options-php-output.ace-enabled.
    // Devel PHP console.
    $('textarea#edit-code.ace-enabled').show(function() {
      ace.require("ace/ext/language_tools");
      ace.require("ace/mode/php");

      var textarea = $(this);
      id = textarea.attr('id');
      var ace_pre_id = id + '-ace';

      textarea.before('<div id="' + ace_pre_id + '"></div>');

      var editor = ace.edit(ace_pre_id);
      var editorSettings = $.extend({}, Drupal.settings.ace_editor.admin);

      $('#' + ace_pre_id).css('height','30em');

      // Set ace configuration options.
      $('#' + ace_pre_id).css('font-size', editorSettings['fontsize']);
      editor.setTheme("ace/theme/" + editorSettings.theme);
      editor.getSession().setMode("ace/mode/php");
      editor.setShowPrintMargin(editorSettings['printmargin']);
      editor.setHighlightActiveLine(editorSettings['linehighlighting']);
      editor.getSession().setFoldStyle(editorSettings['codefolding']);
      editor.renderer.setHScrollBarAlwaysVisible(false);
      // Enable options and snippets.
      editor.setOptions({
        enableBasicAutocompletion: true,
        enableLiveAutocompletion: false,
        enableSnippets: true,
        tabSize: editorSettings['tabsize'],
        useSoftTabs: true
      });

      // Restore the textarea value.
      editor.getSession().setValue(textarea.val());
      editor.focus();

      editor.getSession().on('change', function(){
        textarea.val(editor.getSession().getValue());
      });

      textarea.hide();

    });
    // End Devel PHP console.
  }
};
})(jQuery);
